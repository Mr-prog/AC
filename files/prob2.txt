  section_data
  max_limit $ 4000000 
  fib1 $ 1
  fib2 $ 2
  zero $ 0
  section_program
    LOADMEMORY fib1 R6    ; R6 = 1 (F(n-2))
    LOADMEMORY fib2 R7    ; R7 = 2 (F(n-1))
    LOADMEMORY fib2 R15
    LOADMEMORY zero R14
    LOADMEMORY max_limit R9
    ADD R0 R14 R8         
  fib_loop:
    SUB R9 R7 R10            
    JZ end_program           
    MOD R7 R15 R10            
    JZ is_even               
  next_fib:
    ADD R6 R7 R11            
    ADD R7 R0 R6            
    ADD R11 R0 R7            
    JUMP fib_loop            
  is_even:
    ADD R7 R8 R12
    ADD R12 R0 R8
    JUMP next_fib            
  end_program:
    WRITE_PORT R12            
    NOP                      