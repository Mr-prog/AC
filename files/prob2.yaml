; fib_even_sum.asm
; Вычислить сумму чётных чисел Фибоначчи, не превышающих 4 млн

start:
    LOADI 1, R1                     ; Первое число Фибоначчи
    LOADI 2, R2                     ; Второе число Фибоначчи
    LOADI 0, R3                     ; Сумма чётных чисел
    LOADI 4000000, R4               ; Предел вычислений

fib_loop:
    ADD R1, R2, R5                  ; Вычислить следующее число Фибоначчи
    MOV R2, R1                      ; Обновить R1
    MOV R5, R2                      ; Обновить R2

    MOD R5, 2, R6                   ; Проверка чётности
    JZ R6, add_to_sum               ; Если число чётное, перейти к добавлению

check_limit:
    SUB R4, R5, R7                  ; Проверка, не превышает ли число лимит
    JZ R7, end                      ; Если превышает, завершить программу
    JUMP fib_loop                   ; Иначе продолжить цикл

add_to_sum:
    ADD R3, R5, R3                  ; Добавить чётное число к сумме
    JUMP check_limit                ; Вернуться к проверке предела

end:
    WRITE_PORT R3, POUT             ; Вывести сумму на экран
    NOP                             ; Завершение программы
