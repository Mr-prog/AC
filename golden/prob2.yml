in_source: |-
  CONST Fib1, 1
  CONST Fib2, 2
  LOADMEMORY Fib1, R1              ; Первое число Фибоначчи
  LOADMEMORY Fib2, R2              ; Второе число Фибоначчи
  LOADMEMORY Fib2, R3                     ; Сумма чётных чисел
  LOADI 4000000, R4               ; Предел вычислений
  ADD R1, R2, R5                  ; Вычислить следующее число Фибоначчи
  fib_loop:
    MOV R2, R1                      ; Обновить R1
    MOV R5, R2                      ; Обновить R2
    ADD R1, R2, R5                  ; Вычислить следующее число Фибоначчи
    MOD R5, 2, R6                   ; Проверка чётности
    JZ R6, add_to_sum               ; Если число чётное, перейти к добавлению
  check_limit:
    SUB R4, R5, R7                  ; Проверка, не превышает ли число лимит
    JZ R7, end                      ; Если превышает, завершить программу
    JUMP fib_loop                   ; Иначе продолжить цикл
  add_to_sum:
    ADD R3, R5, R3                  ; Добавить чётное число к сумме
    JUMP check_limit                ; Вернуться к проверке предела
  end:
    WRITE_PORT R3            ; Вывести сумму на экран
    NOP                             ; Завершение программы
in_stdin: ""
out_code: |
  opcode=LOADI operands_type=TWO operands=[1, 1]
  opcode=LOADI operands_type=TWO operands=[2, 2]
  opcode=LOADI operands_type=TWO operands=[0, 3]
  opcode=LOADI operands_type=TWO operands=[4000000, 4]
  opcode=ADD operands_type=THREE operands=[1, 2, 5]
  opcode=JZ operands_type=TWO operands=[5, 13]
  opcode=MOD operands_type=THREE operands=[5, 2, 6]
  opcode=JZ operands_type=TWO operands=[6, 11]
  opcode=MOV operands_type=TWO operands=[2, 1]
  opcode=MOV operands_type=TWO operands=[5, 2]
  opcode=JUMP operands_type=ONE operands=[5]
  opcode=ADD operands_type=THREE operands=[3, 5, 3]
  opcode=JUMP operands_type=ONE operands=[9]
  opcode=WRITE_PORT operands_type=ONE operands=[3]
  opcode=NOP operands_type=NONE operands=[]
out_stdout: "4613732"
out_log: |
  DEBUG:translator:Start translation...
  DEBUG:machine:Running program...
